### PR TITLE
ping: allow run as setuid-root

### PR DESCRIPTION
**Pull request by arjun024**
The iputils-ping version present in the cflinuxfs4 stack does not have the setuid bit set (like that one in cflinuxfs3). The installer checks for LIBCAP[1], and if available sets file capability cap_net_raw+p and do not set the setuid bit - this is happening in our case. However, when ping is run in the CF application container created by garden (vcap user), it fails with "operation not permitted" error. Setuid-rooting works around this.

Tested out by deploying a bosh release and running ping from cf-sshing into the app container.

### More details

tldr: The real root cause is still undetermined.

The following runs fine:
```
$ docker run --user vcap:vcap cloudfoundry/cflinuxfs4 sh -c "ping www.cloudfoundry.org -w1"
```

and you can see the file capabilities of ping
```
$ docker run --user vcap:vcap cloudfoundry/cflinuxfs4 sh -c "getcap /usr/bin/ping"
/usr/bin/ping cap_net_raw=ep
```

but when you run it vai `cf ssh cflinuxfs4-app`:
```
$ /usr/sbin/getcap $(command -v ping)
<no response>
```

The ping program is capabilities aware - it explicitly adds and removes capabilities/effective bit in code ([link](https://github.com/iputils/iputils/blob/8d1acb6d95731ce1b2c7ee1c361bccfaa0918487/ping/ping.c#L585-L614)). This would error if it does not have the permission to gain that desired capability.

On a discussion thread with the Garden team ([link](https://cloudfoundry.slack.com/archives/C033RE5D6/p1694113339696739)), they believe the container should have CAP_NET_RAW, and the possible culpurit could be another undetermined system setting.

It's also possible that since file capabilities are stored as xattr attributes ([man](https://man7.org/linux/man-pages/man7/capabilities.7.html)), and the docker image is "docker-exported" to make it into a bosh-release blob, the export doesn't preserve the attributes!!

1. https://github.com/iputils/iputils/blob/8d1acb6d95731ce1b2c7ee1c361bccfaa0918487/meson.build#L161-L168

### CODE REVIEW

**Review by robdimsdale — APPROVED** [review]
Thank you so much for the deep dive!

I think this proposed solution - `setuid` bit - is an acceptable solution to the problem that `ping` doesn't work out of the box in cflinuxfs4. I don't love it, however, because it _should_ work without it. Following along in the slack chat you linked with the Garden team, I feel like there might be some issue with the capabilities or maybe the `net.ipv4.ping_group_range` system config of the cell or app container.

I think I'm comfortable enough with this to approve and merge, but I can't shake the concern that we're adding a vector for a root-permissions escalation here. The reason I'm ok with this risk is that `ping` is a really common open source program with lots of eyes on the source code. So if there were to be an escalation, I think it would be noticed and fixed quickly.

I will approve but not merge this, to allow you to merge it when you feel like you've exhausted the avenues of investigation.

**Comment by arjun024** [comment]
Update: It looks like the real issue could be in the preservation of file capabilities (as xattr attributes) while the rootfs is tarred and untarred in both the stack-release and in garden. In the stack bosh release, we can preserve xattr attributes [with this diff](https://github.com/cloudfoundry/cflinuxfs4-release/pull/4/files), but garden also has to troubleshoot where they use tar (nstar?) to extract the rootfs and mount the filesystem. ([discussion in #garden](https://cloudfoundry.slack.com/archives/C033RE5D6/p1694200819132979?thread_ts=1694113339.696739&cid=C033RE5D6))

For the moment, we will go ahead with setuid until we have more info from the garden on a better course of action

**Pull request merged**
